Enable custom root directory of build commands

MilesCranmer opened the pull request:
Edit: This is lower priority than #350 and #351 

---

This adds the `-p, --project <root folder>` CLI option so that you can run `showyourwork` from other directories.

I didn't want to break anything so the change is extremely simple: it just changes directories, runs the regular showyourwork commands, then changes back.

@dfm would you be willing to review this?

Fixes #347

MilesCranmer committed:
Add project flag to CLI

dfm commented:
I'm up for including something like this, but I don't love this specific implementation since you'll end up in the wrong directory if anything goes wrong during the `showyourwork` run. I'd recommend something like the following context manager:

```python
import os
from contextlib import contextmanager

@contextmanager
def in_project(path):
    old_dir = os.getcwd()
    os.chdir(path)
    yield
    os.chdir(old_dir)
```

which you could then use as

```python
with in_project(project):
    ensure_top_level()
    commands....
```

to make sure that we always end up chdiring back.

MilesCranmer committed:
Refactor `--project` to use contextmanager

MilesCranmer committed:
Refactor click options to avoid repetition

MilesCranmer commented:
Okay I added the context manager, great idea!

~~Also, I refactored the click options which are used multiple times so there wasn't as much repetition. For example:~~

```diff
- @click.option(
-     "-c",
-     "--cores",
-     default="1",
-     help="Number of cores to use; passed to snakemake.",
- )
+ @takes_cores
```

~~Let me know what you think.~~

MilesCranmer committed:
Refactor click options to be set by common dicts

MilesCranmer commented:
Sorry, am being a perfectionist... I refactored the common options again to be like this instead:

```py
@click.option(*cores_option.args, **cores_options.kwargs)
@click.option(*conda_frontend_option.args, **conda_frontend_option.kwargs)
```

etc. So at least with this it still is obvious that it's a `click.option` rather than some crazy internal macro.

dfm commented:
LGTM - thank you!

MilesCranmer commented:
Thanks for merging!

Do you think this, #350, and #351, could all be released in a new version of SYW? Then I can give a tag instead of the SHA.

dfm commented:
Will do!
